sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala

```diff
@@ -255,41 +260,148 @@ case class HashAggregateExec(
-    val updateExpr = aggregateExpressions.flatMap { e =>
+    val updateExprs = aggregateExpressions.map { e =>
       e.mode match {
         case Partial | Complete =>
           e.aggregateFunction.asInstanceOf[DeclarativeAggregate].updateExpressions
         case PartialMerge | Final =>
           e.aggregateFunction.asInstanceOf[DeclarativeAggregate].mergeExpressions
       }
     }
-    ctx.currentVars = bufVars ++ input
-    val boundUpdateExpr = bindReferences(updateExpr, inputAttrs)
-    val subExprs = ctx.subexpressionEliminationForWholeStageCodegen(boundUpdateExpr)
+    ctx.currentVars = bufVars.flatten ++ input
+    val boundUpdateExprs = updateExprs.map { updateExprsInAgg =>
+      updateExprsInAgg.map(BindReferences.bindReference(_, inputAttrs))
+    }
+    val subExprs = ctx.subexpressionEliminationForWholeStageCodegen(boundUpdateExprs.flatten)
     val effectiveCodes = subExprs.codes.mkString("\n")
-    val aggVals = ctx.withSubExprEliminationExprs(subExprs.states) {
-      boundUpdateExpr.map(_.genCode(ctx))
+    val aggVals = boundUpdateExprs.map { boundUpdateExprsInAgg =>
+      ctx.withSubExprEliminationExprs(subExprs.states) {
+        boundUpdateExprsInAgg.map(_.genCode(ctx))
+      }
     }
-    // aggregate buffer should be updated atomic
-    val updates = aggVals.zipWithIndex.map { case (ev, i) =>
+
+    lazy val nonSplitAggCode = {
+      // aggregate buffer should be updated atomically
+      val updates = aggVals.flatten.zip(bufVars.flatten).map { case (ev, bufVar) =>
+        s"""
+           | ${bufVar.isNull} = ${ev.isNull};
+           | ${bufVar.value} = ${ev.value};
+         """.stripMargin
+      }
       s"""
-         | ${bufVars(i).isNull} = ${ev.isNull};
-         | ${bufVars(i).value} = ${ev.value};
+         | // do aggregate
+         | // common sub-expressions
+         | $effectiveCodes
+         | // evaluate aggregate functions
+         | ${evaluateVariables(aggVals.flatten)}
+         | // update aggregation buffers
+         | ${updates.mkString("\n").trim}
        """.stripMargin
     }
-    s"""
-       | // do aggregate
-       | // common sub-expressions
-       | $effectiveCodes
-       | // evaluate aggregate function
-       | ${evaluateVariables(aggVals)}
-       | // update aggregation buffer
-       | ${updates.mkString("\n").trim}
-     """.stripMargin
+
+    if (conf.codegenSplitAggregateFunc) {
+      val splitAggCode = splitAggregateExpressions(
+        ctx = ctx,
+        aggNames = functions.map(_.prettyName),
+        aggExprs = boundUpdateExprs,
+        makeSplitAggFunctions = {
+          aggVals.zip(bufVars).map { case (aggValsInAgg, bufVarsInAgg) =>
+            // All the update code for aggregation buffers should be placed in the end
+            // of each aggregation function code.
+            val updates = aggValsInAgg.zip(bufVarsInAgg).map { case (ev, bufVar) =>
+              s"""
+                 | ${bufVar.isNull} = ${ev.isNull};
```

Comment:
nit: can we remove the extra space after `|` for the new code? it doesn't help to reduce the diff.

Reply:
removed